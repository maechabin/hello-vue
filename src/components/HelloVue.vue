<template>
  <div>
    <h1>Hello {{ value | convertUpperCase }}</h1>
    <input type="text" v-bind:value="inputValue" v-on:input="handleInput($event)" />
    <button v-on:click="handleClick()" v-bind:disabled="isDisabled">button</button>
  </div>
</template>

<script>
export default {
  /** コンポーネント名 */
  name: 'hello-vue',

  /** 親コンポーネントから受け取るデータ */
  props: ['val'],

  /** コンポーネントに属するデータ（状態） */
  data() {
    return {
      value: this.val,
      inputValue: '',
    };
  },

  /**
   * lifycycle hooks
   * mountedはコンポーネントがマウントされる直前に呼ばれる
   * */
  mounted() {
    console.log('mounted');
  },

  /** いわゆるGet/Setアクセサ */
  computed: {
    isDisabled() {
      return this.inputValue === '';
    },
  },

  /** コンポーネント内で使用するメソッド */
  methods: {
    handleInput($event) {
      this.inputValue = $event.target.value;
    },
    handleClick() {
      if (this.inputValue === '') {
        return;
      }
      this.value = this.inputValue;
      this.inputValue = '';

      /** カスタムイベントのemit */
      this.$emit('handle-click', this.value);
    },
  },

  /** dataの更新を監視し、変更時に呼ばれるメソッド */
  watch: {
    value(newValue, oldValue) {
      console.log(`watch: ${newValue}, ${oldValue}`);
    },
  },

  /** テンプレートの出力の際に任意のフォーマットに変換するためのメソッド */
  filters: {
    convertUpperCase(value) {
      if (!value) return;
      return value.toUpperCase();
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
</style>
