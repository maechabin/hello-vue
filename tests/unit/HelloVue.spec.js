import { shallowMount } from '@vue/test-utils';
import HelloVue from '@/components/HelloVue.vue';

describe('HelloVue.vue', () => {
  let wrapper;

  it('props', () => {
    const val = 'Vue';
    wrapper = shallowMount(HelloVue, {
      propsData: { val },
    });

    expect(wrapper.props().val).toBe(val);
    expect(wrapper.text()).toMatch(`Hello VUE`);
  });

  it('dom', () => {
    wrapper = shallowMount(HelloVue);

    expect(wrapper.contains('h1')).toBeTruthy();
    expect(wrapper.contains('input')).toBeTruthy();
    expect(wrapper.contains('button')).toBeTruthy();
  });

  it('input: v-bind:value', () => {
    wrapper = shallowMount(HelloVue);
    wrapper.setData({ inputValue: 'AAA' });
    expect(wrapper.find('input').element.value).toBe('AAA');
  });

  it('button: v-bind:disabled', () => {
    wrapper = shallowMount(HelloVue);
    wrapper.setData({ inputValue: '' });
    expect(wrapper.find('button').element.disabled).toBeTruthy();
  });

  describe('event', () => {
    beforeEach(() => {
      wrapper = shallowMount(HelloVue);
    });

    it('should call handleInput', () => {
      const spy = jest.spyOn(wrapper.vm, 'handleInput');
      wrapper.find('input').trigger('input');
      expect(spy).toHaveBeenCalled();
    });

    it('should call hancleClick', () => {
      const spy = jest.spyOn(wrapper.vm, 'handleClick');
      wrapper.setData({ inputValue: 'AAA' });
      wrapper.find('button').trigger('click');

      expect(spy).toHaveBeenCalled();
    });

    it('should not call handleClick', () => {
      wrapper = shallowMount(HelloVue);
      const spy = jest.spyOn(wrapper.vm, 'handleClick');
      wrapper.setData({ inputValue: '' });
      wrapper.find('button').trigger('click');

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('mounted', () => {
    it('console.log', () => {
      const spy = jest.spyOn(console, 'log');
      shallowMount(HelloVue);
      expect(spy).toHaveBeenCalled();
    });
  });

  describe('computed: isDisabled', () => {
    beforeEach(() => {
      wrapper = shallowMount(HelloVue);
    });

    it('sholud be true', () => {
      wrapper.setData({ inputValue: '' });
      const disabled = wrapper.vm.isDisabled;

      expect(disabled).toBeTruthy();
    });

    it('sholud be false', () => {
      wrapper.setData({ inputValue: 'AAA' });
      const disabled = wrapper.vm.isDisabled;

      expect(disabled).toBeFalsy();
    });
  });

  describe('method', () => {
    beforeEach(() => {
      wrapper = shallowMount(HelloVue);
    });

    it('handleInput', () => {
      const event = {
        target: { value: 'AAA' },
      };
      wrapper.vm.handleInput(event);
      expect(wrapper.vm.inputValue).toBe('AAA');
    });

    it('handleClick', () => {
      wrapper.setData({ inputValue: 'AAA' });
      const spy = jest.spyOn(wrapper.vm, '$emit');
      wrapper.vm.handleClick();

      expect(wrapper.vm.value).toBe('AAA');
      expect(wrapper.vm.inputValue).toBe('');
      expect(spy).toHaveBeenCalledWith('handle-click', 'AAA');
    });
  });

  describe('watch', () => {
    beforeEach(() => {
      wrapper = shallowMount(HelloVue);
      wrapper.setData({ value: 'AAA' });
    });

    it('should call value()', () => {
      const spy = jest.spyOn(console, 'log');
      wrapper.setData({ value: 'BBB' });

      expect(spy).toHaveBeenCalledWith('watch: BBB, AAA');
    });
  });

  describe('filter', () => {
    it('convertUpperCase', () => {
      wrapper = shallowMount(HelloVue);
      wrapper.setData({ value: 'AAA' });
      const received = wrapper.find('h1').text();
      expect(received).toBe('Hello AAA');
    });
  });

  describe('snapshot', () => {
    it('HelloVue', () => {
      wrapper = shallowMount(HelloVue);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
