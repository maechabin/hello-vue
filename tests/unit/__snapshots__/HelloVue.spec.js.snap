// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloVue.vue snapshot HelloVue 1`] = `
<div>
  <h1>Hello </h1>
  <input type="text">
  <button disabled="disabled">button</button>
</div>
`;
